File: modules/dictionary-audio-sync/src/DictionaryAudioSyncModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import type {
  DictionaryAudioSyncModuleEvents,
  DictionaryAudioSyncStatus,
} from './DictionaryAudioSync.types';

declare class DictionaryAudioSyncModule extends NativeModule<DictionaryAudioSyncModuleEvents> {
  startSync(baseUrl: string, allowCellular: boolean): Promise<DictionaryAudioSyncStatus>;
  pauseSync(): Promise<DictionaryAudioSyncStatus>;
  resumeSync(allowCellular: boolean): Promise<DictionaryAudioSyncStatus>;
  cancelSync(): Promise<DictionaryAudioSyncStatus>;
  clearAll(): Promise<DictionaryAudioSyncStatus>;
  deleteEntry(entryId: string): Promise<DictionaryAudioSyncStatus>;
  getStatus(): Promise<DictionaryAudioSyncStatus>;
  getLocalUri(entryId: string): Promise<string | null>;
}

export default requireNativeModule<DictionaryAudioSyncModule>('DictionaryAudioSync');

File: modules/dictionary-audio-sync/ios/DictionaryAudioSyncModule.swift
import ExpoModulesCore

public final class DictionaryAudioSyncModule: Module {
  private let manager = DictionaryAudioSyncManager.shared

  public func definition() -> ModuleDefinition {
    Name("DictionaryAudioSync")
    Events("onStateChanged", "onProgress")

    OnCreate {
      manager.eventSink = { [weak self] name, body in self?.sendEvent(name, body) }
    }
    OnDestroy { manager.eventSink = nil }

    AsyncFunction("startSync") { (baseURL: String, allowCellular: Bool) in
      try manager.startSync(baseURL: baseURL, allowCellular: allowCellular)
    }
    AsyncFunction("pauseSync") { try manager.pauseSync() }
    AsyncFunction("resumeSync") { (allowCellular: Bool) in try manager.resumeSync(allowCellular: allowCellular) }
    AsyncFunction("cancelSync") { try manager.cancelSync() }
    AsyncFunction("clearAll") { try manager.clearAll() }
    AsyncFunction("deleteEntry") { (entryID: String) in try manager.deleteEntry(entryID) }
    AsyncFunction("getStatus") { manager.status() }
    AsyncFunction("getLocalUri") { (entryID: String) in manager.localURI(entryID: entryID) }
  }
}
